lib/notifications/navigate.dart

```diff
@@ -0,0 +1,320 @@
+  /// Provides the route to open if the app was launched through a tap on
+  /// a notification.
+  ///
+  /// Returns null if app launch wasn't triggered by a notification, or if
+  /// an error occurs while determining the route for the notification.
+  /// In the latter case an error dialog is also shown.
+  ///
+  /// The context argument is used to look up the [Navigator], which is used
+  /// to show an error dialog if there is a failure.
+  AccountRoute<void>? routeForNotificationFromLaunch({required BuildContext context}) {
+    assert(defaultTargetPlatform == TargetPlatform.iOS);
+    final data = _notifDataFromLaunch;
+    if (data == null) return null;
+    assert(debugLog('opened notif: ${jsonEncode(data.payload)}'));
+
+    final notifNavData = _tryParseIosApnsPayload(context, data.payload);
+    if (notifNavData == null) return null; // TODO(log)
+
+    return routeForNotification(context, notifNavData);
+  }
+
+  /// Provides the route to open by parsing the notification payload.
+  ///
+  /// Returns null and shows an error dialog if the associated account is not
+  /// found in the global store.
+  static AccountRoute<void>? routeForNotification(
+    BuildContext context,
+    NotificationNavigationData data,
+  ) {
+    final globalStore = GlobalStoreWidget.of(context);
+
+    final account = globalStore.accounts.firstWhereOrNull(
+      (account) => account.realmUrl.origin == data.realmUrl.origin
+                && account.userId == data.userId);
+    if (account == null) { // TODO(log)
+      final zulipLocalizations = ZulipLocalizations.of(context);
+      showErrorDialog(context: context,
+        title: zulipLocalizations.errorNotificationOpenTitle,
+        message: zulipLocalizations.errorNotificationOpenAccountNotFound);
+      return null;
+    }
+
+    return MessageListPage.buildRoute(
+      accountId: account.id,
+      // TODO(#82): Open at specific message, not just conversation
+      narrow: data.narrow);
+  }
+
+  /// Navigates to the [MessageListPage] of the specific conversation
+  /// for the provided payload that was attached while creating the
+  /// notification.
+  Future<void> _navigateForNotification(NotificationTapEvent event) async {
+    assert(debugLog('opened notif: ${jsonEncode(event.payload)}'));
+
+    NavigatorState navigator = await ZulipApp.navigator;
+    final context = navigator.context;
+    assert(context.mounted);
+    if (!context.mounted) return; // TODO(linter): this is impossible as there's no actual async gap, but the use_build_context_synchronously lint doesn't see that
+
+    final notifNavData = _tryParseIosApnsPayload(context, event.payload);
+    if (notifNavData == null) return; // TODO(log)
+    final route = routeForNotification(context, notifNavData);
+    if (route == null) return; // TODO(log)
+
+    // TODO(nav): Better interact with existing nav stack on notif open
+    unawaited(navigator.push(route));
+  }
+
+  NotificationNavigationData? _tryParseIosApnsPayload(
```

Comment:
Can this be static? Seems similar to `tryParseAndroidNotificationUrl` just below it.